File: examples/quickstart-typescript/app/domain.ts
/**
 * Tiny e-commerce domain model for the Stele TypeScript quickstart demo.
 *
 * Stele contracts (contract/main.stele) enforce business rules on these types
 * without requiring changes here. Run `npx stele generate && npx vitest run`
 * to verify.
 */

export interface Item {
  sku: string;
  price: number;
  qty: number;
}

export interface Order {
  id: string;
  total: number;
  items: Item[];
}

export interface User {
  id: string;
  email: string;
  /** "active" | "suspended" | "deleted" */
  status: string;
}

export type UserStatus = "active" | "suspended" | "deleted";

export const ALLOWED_USER_STATUSES: readonly UserStatus[] = ["active", "suspended", "deleted"];

File: examples/quickstart-python/app/domain.py
"""
Tiny e-commerce domain model for the Stele quickstart demo.

Stele contracts (contract/main.stele) enforce business rules on these objects
without requiring any code changes here. Run `npx stele generate && python -m
pytest tests/contract` to verify.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal


@dataclass(frozen=True)
class Item:
    sku: str
    price: Decimal
    qty: int


@dataclass
class Order:
    id: str
    total: Decimal
    items: list[Item] = field(default_factory=list)

    def recalculate_total(self) -> None:
        self.total = sum((item.price * item.qty for item in self.items), Decimal("0"))


@dataclass
class User:
    id: str
    email: str
    status: str  # "active" | "suspended" | "deleted"

    ALLOWED_STATUSES = ("active", "suspended", "deleted")

    def is_active(self) -> bool:
        return self.status == "active"
